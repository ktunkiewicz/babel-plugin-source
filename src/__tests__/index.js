/* global expect */

import path from 'path'
import pluginTester from 'babel-plugin-tester'
import plugin from '../'

const projectRoot = path.join(__dirname, '../../')

expect.addSnapshotSerializer({
  print (val) {
    return val.split(projectRoot).join('<PROJECT_ROOT>/')
  },
  test (val) {
    return typeof val === 'string'
  },
})

const shouldError = code => ({ code, error: true })

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {
    presets: ['react'],
  },
  tests: [
    `
      let MySource
      function ScopedFunction () {
        /* @source MySource */ const foo = 'bar' /* @source MySource */
      }
    `,
    `
      function ScopedFunction () {
        let MySource
        /* @source MySource */ const foo = 'bar' /* @source MySource */
      }
    `,
    `
      let MySource
      /* @source MySource */ const foo = 'bar' /* @source MySource */
    `,
    `
      let MySource
      /* @source MySource */ const foo = 'bar' /* @source MySource */
    `,
    `
      let MySource
      // @source MySource
      const MyComponent = () => (
        console.log('Hello!')
      )
      // @source MySource
    `,
    `
      let MySource
      // @source MySource
      const MyComponent = () => (
        console.log('Hello!')
      )
      // @source MySource
    `,
    `
      let MySource
      let MyOtherSource
      // @source MySource
      const MyComponent = () => (
        console.log('Hello!')
      )
      // @source MySource
      // @source MyOtherSource
      const MyComponent2 = () => (
        console.log('Hello 2!')
      )
      // @source MyOtherSource
    `,
    `
      let MySource
      // @source MySource
      // @source MySource
    `,
    `
      let MySource
      // @source MySource
      // @source MySource
    `,
    `
      let MySource = 'I should be the here'
      /* @source */ const foo = 'bar' /* @source */
    `,
    `
      let MySource = 'I should be the here'
      /* other comment */ const foo = 'bar' /* other comment */
    `,
    shouldError(`
      /* @source MySource */ const foo = 'bar' /* @source MySource */
    `),
    `
      /* other comment */ const foo = 'bar' /* other comment */
    `,
    `
      let MySources
      let MySource2
      let MySource3
      let MySource4

      /* @source MySources */

      if (true) {
        // do this
      }
      const MyComponent = () => (
        /* @source MySource2 */
        console.log('Hello!')
        /* @source MySource2 */

      )
      const AnotherComponent = () => (
        // @source MySource3
        console.log('Hello Again!')
        // @source MySource3
      )

      /* @source MySource4 */ const aVariable = 'Hello!' /* @source MySource4 */

      /* @source MySources */

      const noSource = 'please!'
    `,
    `
      let MySource
      const MyComponent = () => (
        // @source MySource
        console.log('Hello!')
        // @source MySource
      )
    `,
    shouldError(`
      const MyComponent = () => (
        // @source MySource
        console.log('Hello!')
        // @source MySource
      )
    `),
    `
      const MySource = ''
  
      const MyComponent = () => (
        // @source MySource
        console.log('Hello!')
        // @source MySource
      )
    `,
    `
      let MyJSXSource
      const MyComponent = () => (
        // @source MyJSXSource
        <div>Hello!</div>
        // @source MyJSXSource
      )
    `,
    `
      let sourceCode

      console.log(sourceCode)

      export default () => (
        <Sidebar>
          <ChartConfig dataType="time">
            {({ data }) => (
              // @source sourceCode
              <Chart data={data}>
                <Axis primary type="time" position="bottom" />
                <Axis type="linear" position="left" stacked />
                <Series type={Area} />
                <Tooltip />
              </Chart>
              // @source sourceCode
            )}
          </ChartConfig>
          <Code source={sourceCode} />
        </Sidebar>
      )
    `,
    `import React from 'react'

      //
      
      import Sidebar from 'components/Sidebar'
      import ChartConfig from 'components/ChartConfig'
      import Code from 'components/Code'
      
      import { Chart, Axis, Series, Tooltip, Area } from '../../../src'
      
      const sourceCode = ''
      
      export default () => (
        <Sidebar>
          <ChartConfig dataType="time">
            {({ data }) => (
              // @source sourceCode
              <Chart data={data}>
                <Axis primary type="time" position="bottom" />
                <Axis type="linear" position="left" stacked />
                <Series type={Area} />
                <Tooltip />
              </Chart>
              // @source sourceCode
            )}
          </ChartConfig>
          <Code source={sourceCode} />
        </Sidebar>
      )
    `,
  ],
})
