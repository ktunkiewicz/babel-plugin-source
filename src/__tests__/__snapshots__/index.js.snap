// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. source 1`] = `

let MySource
function ScopedFunction () {
  /* @source MySource */ const foo = 'bar' /* @source MySource */
}

      ↓ ↓ ↓ ↓ ↓ ↓

let MySource = 'const foo = \\'bar\\' ';
function ScopedFunction() {
  /* @source MySource */const foo = 'bar'; /* @source MySource */
}

`;

exports[`2. source 1`] = `

function ScopedFunction () {
  let MySource
  /* @source MySource */ const foo = 'bar' /* @source MySource */
}

      ↓ ↓ ↓ ↓ ↓ ↓

function ScopedFunction() {
  let MySource = 'const foo = \\'bar\\' ';
  /* @source MySource */const foo = 'bar'; /* @source MySource */
}

`;

exports[`3. source 1`] = `

let MySource
/* @source MySource */ const foo = 'bar' /* @source MySource */

      ↓ ↓ ↓ ↓ ↓ ↓

let MySource = 'const foo = \\'bar\\' ';
/* @source MySource */const foo = 'bar'; /* @source MySource */

`;

exports[`4. source 1`] = `

let MySource
/* @source MySource */ const foo = 'bar' /* @source MySource */

      ↓ ↓ ↓ ↓ ↓ ↓

let MySource = 'const foo = \\'bar\\' ';
/* @source MySource */const foo = 'bar'; /* @source MySource */

`;

exports[`5. source 1`] = `

let MySource
// @source MySource
const MyComponent = () => (
  console.log('Hello!')
)
// @source MySource

      ↓ ↓ ↓ ↓ ↓ ↓

let MySource = 'const MyComponent = () => (\\n  console.log(\\'Hello!\\')\\n)';
// @source MySource
const MyComponent = () => console.log('Hello!');
// @source MySource

`;

exports[`6. source 1`] = `

let MySource
// @source MySource
const MyComponent = () => (
  console.log('Hello!')
)
// @source MySource

      ↓ ↓ ↓ ↓ ↓ ↓

let MySource = 'const MyComponent = () => (\\n  console.log(\\'Hello!\\')\\n)';
// @source MySource
const MyComponent = () => console.log('Hello!');
// @source MySource

`;

exports[`7. source 1`] = `

let MySource
let MyOtherSource
// @source MySource
const MyComponent = () => (
  console.log('Hello!')
)
// @source MySource
// @source MyOtherSource
const MyComponent2 = () => (
  console.log('Hello 2!')
)
// @source MyOtherSource

      ↓ ↓ ↓ ↓ ↓ ↓

let MySource = 'const MyComponent = () => (\\n  console.log(\\'Hello!\\')\\n)';
let MyOtherSource = 'const MyComponent2 = () => (\\n  console.log(\\'Hello 2!\\')\\n)';
// @source MySource
const MyComponent = () => console.log('Hello!');
// @source MySource
// @source MyOtherSource
const MyComponent2 = () => console.log('Hello 2!');
// @source MyOtherSource

`;

exports[`8. source 1`] = `

let MySource
// @source MySource
// @source MySource

      ↓ ↓ ↓ ↓ ↓ ↓

let MySource = "";
// @source MySource
// @source MySource

`;

exports[`9. source 1`] = `

let MySource
// @source MySource
// @source MySource

      ↓ ↓ ↓ ↓ ↓ ↓

let MySource = "";
// @source MySource
// @source MySource

`;

exports[`10. source 1`] = `

let MySource = 'I should be the here'
/* @source */ const foo = 'bar' /* @source */

      ↓ ↓ ↓ ↓ ↓ ↓

let MySource = 'I should be the here';
/* @source */const foo = 'bar'; /* @source */

`;

exports[`11. source 1`] = `

let MySource = 'I should be the here'
/* other comment */ const foo = 'bar' /* other comment */

      ↓ ↓ ↓ ↓ ↓ ↓

let MySource = 'I should be the here';
/* other comment */const foo = 'bar'; /* other comment */

`;

exports[`12. source 1`] = `

/* @source MySource */ const foo = 'bar' /* @source MySource */

      ↓ ↓ ↓ ↓ ↓ ↓

SyntaxError: unknown: Couldn't find the 'let/var MySource' variable for @source MySource!

`;

exports[`13. source 1`] = `

/* other comment */ const foo = 'bar' /* other comment */

      ↓ ↓ ↓ ↓ ↓ ↓

/* other comment */const foo = 'bar'; /* other comment */

`;

exports[`14. source 1`] = `

let MySources
let MySource2
let MySource3
let MySource4

/* @source MySources */

if (true) {
  // do this
}
const MyComponent = () => (
  /* @source MySource2 */
  console.log('Hello!')
  /* @source MySource2 */

)
const AnotherComponent = () => (
  // @source MySource3
  console.log('Hello Again!')
  // @source MySource3
)

/* @source MySource4 */ const aVariable = 'Hello!' /* @source MySource4 */

/* @source MySources */

const noSource = 'please!'

      ↓ ↓ ↓ ↓ ↓ ↓

let MySources = 'if (true) {\\n  // do this\\n}\\nconst MyComponent = () => (\\n  /* @source MySource2 */\\n  console.log(\\'Hello!\\')\\n  /* @source MySource2 */\\n)\\nconst AnotherComponent = () => (\\n  // @source MySource3\\n  console.log(\\'Hello Again!\\')\\n  // @source MySource3\\n)\\n/* @source MySource4 */ const aVariable = \\'Hello!\\' /* @source MySource4 */';
let MySource2 = 'console.log(\\'Hello!\\')';
let MySource3 = 'console.log(\\'Hello Again!\\')';
let MySource4 = 'const aVariable = \\'Hello!\\' ';

/* @source MySources */

if (true) {
  // do this
}
const MyComponent = () => console.log('Hello!');
const AnotherComponent = () => console.log('Hello Again!');

/* @source MySource4 */const aVariable = 'Hello!'; /* @source MySource4 */

/* @source MySources */

const noSource = 'please!';

`;

exports[`15. source 1`] = `

let MySource
const MyComponent = () => (
  // @source MySource
  console.log('Hello!')
  // @source MySource
)

      ↓ ↓ ↓ ↓ ↓ ↓

let MySource = 'console.log(\\'Hello!\\')';
const MyComponent = () => console.log('Hello!');

`;

exports[`16. source 1`] = `

const MyComponent = () => (
  // @source MySource
  console.log('Hello!')
  // @source MySource
)

      ↓ ↓ ↓ ↓ ↓ ↓

SyntaxError: unknown: Couldn't find the 'let/var MySource' variable for @source MySource!

`;

exports[`17. source 1`] = `

const MySource = ''
  
const MyComponent = () => (
  // @source MySource
  console.log('Hello!')
  // @source MySource
)

      ↓ ↓ ↓ ↓ ↓ ↓

const MySource = 'console.log(\\'Hello!\\')';

const MyComponent = () => console.log('Hello!');

`;

exports[`18. source 1`] = `

let MyJSXSource
const MyComponent = () => (
  // @source MyJSXSource
  <div>Hello!</div>
  // @source MyJSXSource
)

      ↓ ↓ ↓ ↓ ↓ ↓

let MyJSXSource = "<div>Hello!</div>";
const MyComponent = () => React.createElement(
  "div",
  null,
  "Hello!"
);

`;

exports[`19. source 1`] = `

let sourceCode

console.log(sourceCode)

export default () => (
  <Sidebar>
    <ChartConfig dataType="time">
      {({ data }) => (
        // @source sourceCode
        <Chart data={data}>
          <Axis primary type="time" position="bottom" />
          <Axis type="linear" position="left" stacked />
          <Series type={Area} />
          <Tooltip />
        </Chart>
        // @source sourceCode
      )}
    </ChartConfig>
    <Code source={sourceCode} />
  </Sidebar>
)

      ↓ ↓ ↓ ↓ ↓ ↓

let sourceCode = "<Chart data={data}>\\n  <Axis primary type=\\"time\\" position=\\"bottom\\" />\\n  <Axis type=\\"linear\\" position=\\"left\\" stacked />\\n  <Series type={Area} />\\n  <Tooltip />\\n</Chart>";

console.log(sourceCode);

export default (() => React.createElement(
  Sidebar,
  null,
  React.createElement(
    ChartConfig,
    { dataType: "time" },
    ({ data }) => React.createElement(
      Chart,
      { data: data },
      React.createElement(Axis, { primary: true, type: "time", position: "bottom" }),
      React.createElement(Axis, { type: "linear", position: "left", stacked: true }),
      React.createElement(Series, { type: Area }),
      React.createElement(Tooltip, null)
    )
  ),
  React.createElement(Code, { source: sourceCode })
));

`;

exports[`20. source 1`] = `

import React from 'react'

      //
      
      import Sidebar from 'components/Sidebar'
      import ChartConfig from 'components/ChartConfig'
      import Code from 'components/Code'
      
      import { Chart, Axis, Series, Tooltip, Area } from '../../../src'
      
      const sourceCode = ''
      
      export default () => (
        <Sidebar>
          <ChartConfig dataType="time">
            {({ data }) => (
              // @source sourceCode
              <Chart data={data}>
                <Axis primary type="time" position="bottom" />
                <Axis type="linear" position="left" stacked />
                <Series type={Area} />
                <Tooltip />
              </Chart>
              // @source sourceCode
            )}
          </ChartConfig>
          <Code source={sourceCode} />
        </Sidebar>
      )

      ↓ ↓ ↓ ↓ ↓ ↓

import React from 'react';

//

import Sidebar from 'components/Sidebar';
import ChartConfig from 'components/ChartConfig';
import Code from 'components/Code';

import { Chart, Axis, Series, Tooltip, Area } from '../../../src';

const sourceCode = '<Chart data={data}>\\n  <Axis primary type="time" position="bottom" />\\n  <Axis type="linear" position="left" stacked />\\n  <Series type={Area} />\\n  <Tooltip />\\n</Chart>';

export default (() => React.createElement(
  Sidebar,
  null,
  React.createElement(
    ChartConfig,
    { dataType: 'time' },
    ({ data }) => React.createElement(
      Chart,
      { data: data },
      React.createElement(Axis, { primary: true, type: 'time', position: 'bottom' }),
      React.createElement(Axis, { type: 'linear', position: 'left', stacked: true }),
      React.createElement(Series, { type: Area }),
      React.createElement(Tooltip, null)
    )
  ),
  React.createElement(Code, { source: sourceCode })
));

`;
